README
xfce4-verve-plugin-smartbookmark

Here is the dependency information from the original README, which hasn't changed in this fork.

Required packages:
  * exo >= 0.3.1.3
  * xfce4-panel >= 4.3.28
  * libxfce4util >= 4.3.0
  * GLib >= 2.6.4
  * pcre

Optionally required packages:
  * D-BUS >= 0.34


debian/changelog
xfce4-verve-plugin-smartbookmark (1.0.6.2) testing; urgency=low

  * Merge debian packaging changes from xfce4-verve-plugin 1.0.0-2 (debian unstable).

 -- Isaac Schemm <isaacschemm@gmail.com>  Thu, 06 Feb 2014 16:45:13 -0600

xfce4-verve-plugin-smartbookmark (1.0.6.1) precise; urgency=low

  * Fixed bug when initializing plugin label

 -- Isaac Schemm <isaacschemm@gmail.com>  Wed, 23 Jan 2013 22:07:11 -0600

xfce4-verve-plugin-smartbookmark (1.0.6) precise; urgency=low

  * Added new option to redirect commands beinning with a backslash to DuckDuckGo, separate from the existing option for !bang commands
  * Added a label next to the main text entry box (set to an empty string by default)

 -- Isaac Schemm <isaacschemm@gmail.com>  Tue, 22 Jan 2013 20:49:04 -0600

xfce4-verve-plugin-smartbookmark (1.0.5) precise; urgency=low

  * Replaced search engine list with smartbookmark functionality

 -- Isaac Schemm <isaacschemm@gmail.com>  Sat, 19 Jan 2013 20:09:57 -0600

xfce4-verve-plugin-ddg (1.0.4) precise; urgency=low

  * Updated translations from official Xfce Git repo (git.xfce.org/panel-plugins/xfce4-verve-plugin)

 -- Isaac Schemm <isaacschemm@gmail.com>  Fri, 18 Jan 2013 13:22:51 -0600

xfce4-verve-plugin-ddg (1.0.3) precise; urgency=low

  * Fixed use of quotes in search queries (by using g_strconcat and replacing xdg-open with sensible-browser)

 -- Isaac Schemm <isaacschemm@gmail.com>  Tue, 15 May 2012 20:07:20 -0500

xfce4-verve-plugin-ddg (1.0.2.1) precise; urgency=low

  * Added Startpage to the list

 -- Isaac Schemm <isaacschemm@gmail.com>  Sat, 05 May 2012 20:41:47 -0500

xfce4-verve-plugin-ddg (1.0.2) oneiric; urgency=low

  * Uploaded to PPA

 -- Isaac Schemm <isaacschemm@gmail.com>  Tue, 27 Mar 2012 17:50:48 -0500

xfce4-verve-plugin-ddg (1.0.1) unstable; urgency=low

  * Initial package build with DuckDuckGo support
  * Raised max size from 100 to 300

 -- Isaac Schemm <isaacschemm@gmail.com>  Wed, 22 Feb 2012 13:36:18 -0600



panel-plugin/verve.h
/***************************************************************************
 *            verve.h
 *
 *  Copyright  2007  Jannis Pohlmann
 *  jannis@xfce.org
 *  DuckDuckGo support (C) 2012 Isaac Schemm
 *  isaacschemm@gmail.com
 ****************************************************************************/

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
#ifndef __VERVE_H__
#define __VERVE_H__

#include "verve-env.h"
#include "verve-history.h"

void verve_set_bang_setting (gboolean bang);
void verve_set_search_engine (gint engine);

/* Init / Shutdown Verve */
void verve_init (void);
void verve_shutdown (void);

/* Command line methods */
gboolean verve_spawn_command_line (const gchar *cmdline);
gboolean verve_execute (const gchar *input, gboolean terminal);

#endif /* !__VERVE_H__ */

/* vim:set expandtab ts=1 sw=2: */


panel-plugin/verve-dbus-service-infos.h
/* Generated by dbus-binding-tool; do not edit! */


#ifndef __dbus_glib_marshal_verve_dbus_service_MARSHAL_H__
#define __dbus_glib_marshal_verve_dbus_service_MARSHAL_H__

#include	<glib-object.h>

G_BEGIN_DECLS

#ifdef G_ENABLE_DEBUG
#define g_marshal_value_peek_boolean(v)  g_value_get_boolean (v)
#define g_marshal_value_peek_char(v)     g_value_get_char (v)
#define g_marshal_value_peek_uchar(v)    g_value_get_uchar (v)
#define g_marshal_value_peek_int(v)      g_value_get_int (v)
#define g_marshal_value_peek_uint(v)     g_value_get_uint (v)
#define g_marshal_value_peek_long(v)     g_value_get_long (v)
#define g_marshal_value_peek_ulong(v)    g_value_get_ulong (v)
#define g_marshal_value_peek_int64(v)    g_value_get_int64 (v)
#define g_marshal_value_peek_uint64(v)   g_value_get_uint64 (v)
#define g_marshal_value_peek_enum(v)     g_value_get_enum (v)
#define g_marshal_value_peek_flags(v)    g_value_get_flags (v)
#define g_marshal_value_peek_float(v)    g_value_get_float (v)
#define g_marshal_value_peek_double(v)   g_value_get_double (v)
#define g_marshal_value_peek_string(v)   (char*) g_value_get_string (v)
#define g_marshal_value_peek_param(v)    g_value_get_param (v)
#define g_marshal_value_peek_boxed(v)    g_value_get_boxed (v)
#define g_marshal_value_peek_pointer(v)  g_value_get_pointer (v)
#define g_marshal_value_peek_object(v)   g_value_get_object (v)
#define g_marshal_value_peek_variant(v)  g_value_get_variant (v)
#else /* !G_ENABLE_DEBUG */
/* WARNING: This code accesses GValues directly, which is UNSUPPORTED API.
 *          Do not access GValues directly in your code. Instead, use the
 *          g_value_get_*() functions
 */
#define g_marshal_value_peek_boolean(v)  (v)->data[0].v_int
#define g_marshal_value_peek_char(v)     (v)->data[0].v_int
#define g_marshal_value_peek_uchar(v)    (v)->data[0].v_uint
#define g_marshal_value_peek_int(v)      (v)->data[0].v_int
#define g_marshal_value_peek_uint(v)     (v)->data[0].v_uint
#define g_marshal_value_peek_long(v)     (v)->data[0].v_long
#define g_marshal_value_peek_ulong(v)    (v)->data[0].v_ulong
#define g_marshal_value_peek_int64(v)    (v)->data[0].v_int64
#define g_marshal_value_peek_uint64(v)   (v)->data[0].v_uint64
#define g_marshal_value_peek_enum(v)     (v)->data[0].v_long
#define g_marshal_value_peek_flags(v)    (v)->data[0].v_ulong
#define g_marshal_value_peek_float(v)    (v)->data[0].v_float
#define g_marshal_value_peek_double(v)   (v)->data[0].v_double
#define g_marshal_value_peek_string(v)   (v)->data[0].v_pointer
#define g_marshal_value_peek_param(v)    (v)->data[0].v_pointer
#define g_marshal_value_peek_boxed(v)    (v)->data[0].v_pointer
#define g_marshal_value_peek_pointer(v)  (v)->data[0].v_pointer
#define g_marshal_value_peek_object(v)   (v)->data[0].v_pointer
#define g_marshal_value_peek_variant(v)  (v)->data[0].v_pointer
#endif /* !G_ENABLE_DEBUG */


/* BOOLEAN:POINTER */
extern void dbus_glib_marshal_verve_dbus_service_BOOLEAN__POINTER (GClosure     *closure,
                                                                   GValue       *return_value,
                                                                   guint         n_param_values,
                                                                   const GValue *param_values,
                                                                   gpointer      invocation_hint,
                                                                   gpointer      marshal_data);
void
dbus_glib_marshal_verve_dbus_service_BOOLEAN__POINTER (GClosure     *closure,
                                                       GValue       *return_value G_GNUC_UNUSED,
                                                       guint         n_param_values,
                                                       const GValue *param_values,
                                                       gpointer      invocation_hint G_GNUC_UNUSED,
                                                       gpointer      marshal_data)
{
  typedef gboolean (*GMarshalFunc_BOOLEAN__POINTER) (gpointer     data1,
                                                     gpointer     arg_1,
                                                     gpointer     data2);
  register GMarshalFunc_BOOLEAN__POINTER callback;
  register GCClosure *cc = (GCClosure*) closure;
  register gpointer data1, data2;
  gboolean v_return;

  g_return_if_fail (return_value != NULL);
  g_return_if_fail (n_param_values == 2);

  if (G_CCLOSURE_SWAP_DATA (closure))
    {
      data1 = closure->data;
      data2 = g_value_peek_pointer (param_values + 0);
    }
  else
    {
      data1 = g_value_peek_pointer (param_values + 0);
      data2 = closure->data;
    }
  callback = (GMarshalFunc_BOOLEAN__POINTER) (marshal_data ? marshal_data : cc->callback);

  v_return = callback (data1,
                       g_marshal_value_peek_pointer (param_values + 1),
                       data2);

  g_value_set_boolean (return_value, v_return);
}

/* BOOLEAN:STRING,STRING,POINTER */
extern void dbus_glib_marshal_verve_dbus_service_BOOLEAN__STRING_STRING_POINTER (GClosure     *closure,
                                                                                 GValue       *return_value,
                                                                                 guint         n_param_values,
                                                                                 const GValue *param_values,
                                                                                 gpointer      invocation_hint,
                                                                                 gpointer      marshal_data);
void
dbus_glib_marshal_verve_dbus_service_BOOLEAN__STRING_STRING_POINTER (GClosure     *closure,
                                                                     GValue       *return_value G_GNUC_UNUSED,
                                                                     guint         n_param_values,
                                                                     const GValue *param_values,
                                                                     gpointer      invocation_hint G_GNUC_UNUSED,
                                                                     gpointer      marshal_data)
{
  typedef gboolean (*GMarshalFunc_BOOLEAN__STRING_STRING_POINTER) (gpointer     data1,
                                                                   gpointer     arg_1,
                                                                   gpointer     arg_2,
                                                                   gpointer     arg_3,
                                                                   gpointer     data2);
  register GMarshalFunc_BOOLEAN__STRING_STRING_POINTER callback;
  register GCClosure *cc = (GCClosure*) closure;
  register gpointer data1, data2;
  gboolean v_return;

  g_return_if_fail (return_value != NULL);
  g_return_if_fail (n_param_values == 4);

  if (G_CCLOSURE_SWAP_DATA (closure))
    {
      data1 = closure->data;
      data2 = g_value_peek_pointer (param_values + 0);
    }
  else
    {
      data1 = g_value_peek_pointer (param_values + 0);
      data2 = closure->data;
    }
  callback = (GMarshalFunc_BOOLEAN__STRING_STRING_POINTER) (marshal_data ? marshal_data : cc->callback);

  v_return = callback (data1,
                       g_marshal_value_peek_string (param_values + 1),
                       g_marshal_value_peek_string (param_values + 2),
                       g_marshal_value_peek_pointer (param_values + 3),
                       data2);

  g_value_set_boolean (return_value, v_return);
}

G_END_DECLS

#endif /* __dbus_glib_marshal_verve_dbus_service_MARSHAL_H__ */

#include <dbus/dbus-glib.h>
static const DBusGMethodInfo dbus_glib_verve_dbus_service_methods[] = {
  { (GCallback) verve_dbus_service_open_dialog, dbus_glib_marshal_verve_dbus_service_BOOLEAN__STRING_STRING_POINTER, 0 },
  { (GCallback) verve_dbus_service_open_dialog, dbus_glib_marshal_verve_dbus_service_BOOLEAN__STRING_STRING_POINTER, 53 },
  { (GCallback) verve_dbus_service_grab_focus, dbus_glib_marshal_verve_dbus_service_BOOLEAN__POINTER, 102 },
};

const DBusGObjectInfo dbus_glib_verve_dbus_service_object_info = {  1,
  dbus_glib_verve_dbus_service_methods,
  3,
"org.xfce.RunDialog\0OpenDialog\0S\0dir\0I\0s\0display\0I\0s\0\0org.xfce.Verve\0OpenDialog\0S\0dir\0I\0s\0display\0I\0s\0\0org.xfce.Verve\0GrabFocus\0S\0\0\0",
"\0",
"\0"
};



panel-plugin/verve.c
/***************************************************************************
 *            verve.c
 *
 *  Copyright  2006-2007  Jannis Pohlmann
 *  jannis@xfce.org
 *  DuckDuckGo support (C) 2012-13 Isaac Schemm
 *  isaacschemm@gmail.com
 ****************************************************************************/

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
#include <pcre.h>

#include <glib-object.h>

#include <libxfcegui4/libxfcegui4.h>

#include "verve.h"
#include "verve-env.h"
#include "verve-history.h"



static gboolean verve_is_url       (const gchar *str);
static gboolean verve_is_email     (const gchar *str);
static gboolean verve_is_directory (const gchar *str);



/* URL/eMail matching patterns */
#define USERCHARS   "-A-Za-z0-9"
#define PASSCHARS   "-A-Za-z0-9,?;.:/!%$^*&~\"#'"
#define HOSTCHARS   "-A-Za-z0-9"
#define USER        "[" USERCHARS "]+(:["PASSCHARS "]+)?"
#define MATCH_URL1  "^((file|https?|ftps?)://(" USER "@)?)[" HOSTCHARS ".]+(:[0-9]+)?" \
                    "(/[-A-Za-z0-9_$.+!*(),;:@&=?/~#%]*[^]'.}>) \t\r\n,\\\"])?/?$"
#define MATCH_URL2  "^(www|ftp)[" HOSTCHARS "]*\\.[" HOSTCHARS ".]+(:[0-9]+)?" \
                    "(/[-A-Za-z0-9_$.+!*(),;:@&=?/~#%]*[^]'.}>) \t\r\n,\\\"])?/?$"
#define MATCH_EMAIL "^(mailto:)?[a-z0-9][a-z0-9.-]*@[a-z0-9][a-z0-9-]*(\\.[a-z0-9][a-z0-9-]*)+$"
#define MATCH_BANG  "^!"
#define MATCH_BACKSLASH "^\\\\"

static gboolean use_bang = TRUE;
static gboolean use_backslash = FALSE;
static gboolean use_smartbookmark = FALSE;
/* URL for smartbookmark */
static gchar* url = "";

void
verve_set_bang_setting (gboolean bang)
{
  use_bang = bang;
}

void
verve_set_backslash_setting (gboolean backslash)
{
  use_backslash = backslash;
}

void
verve_set_smartbookmark_setting (gboolean smartbookmark)
{
  use_smartbookmark = smartbookmark;
}

void
verve_set_url (gchar* engine)
{
  url = g_strdup(engine);
}

/*********************************************************************
 *
 * Initialize/shutdown Verve
 * -------------------------
 *
 * The execution of these functions is necessary before and after 
 * your program makes use of Verve.
 *
 *********************************************************************/

void 
verve_init (void)
{
  /* Init history database */
  verve_history_init ();
}



void
verve_shutdown (void)
{
  /* Free history database */
  verve_history_shutdown ();

  /* Shutdown environment */
  verve_env_shutdown ();
}



/*********************************************************************
 *
 * Verve command line execution function
 * -------------------------------------
 *
 * With the help of this function, shell commands can be executed.
 *
 *********************************************************************/
 
gboolean verve_spawn_command_line (const gchar *cmdline)
{
  gint         argc;
  gchar      **argv;
  gboolean     success;
  GError      *error = NULL;
  const gchar *home_dir;
  GSpawnFlags  flags;
  
  /* Parse command line arguments */
  success = g_shell_parse_argv (cmdline, &argc, &argv, &error);

  /* Return false if command line arguments failed to be arsed */
  if (G_UNLIKELY (error != NULL))
    {
      g_error_free (error);
      return FALSE;
    }

  /* Get user's home directory */
  home_dir = xfce_get_homedir ();

  /* Set up spawn flags */
  flags = G_SPAWN_STDOUT_TO_DEV_NULL;
  flags |= G_SPAWN_STDERR_TO_DEV_NULL;
  flags |= G_SPAWN_SEARCH_PATH;
  
  /* Spawn subprocess */
  success = g_spawn_async (home_dir, argv, NULL, flags, NULL, NULL, NULL, &error);

  /* Return false if subprocess could not be spawned */
  if (G_UNLIKELY (error != NULL)) 
    {
      g_error_free (error);
      return FALSE;
    }

  /* Free command line arguments */
  g_strfreev (argv);
  
  /* Return whether process was spawned successfully */
  return success;
}



/*********************************************************************
 * 
 * Verve main execution method
 * ---------------------------
 *
 * This method should be used whenever you want to run command line 
 * input, be it a URL, an eMail address or a custom command.
 *
 *********************************************************************/

gboolean
verve_execute (const gchar *input, 
               gboolean terminal)
{
  gchar   *command;
  gboolean result = FALSE;
    
  /* Open URLs, eMail addresses and directories using exo-open */
  if (verve_is_url (input) || verve_is_email (input) || verve_is_directory (input))
  {
    /* Build exo-open command */
    command = g_strconcat ("exo-open ", input, NULL);
  }
  else
  {
    if ((use_bang && verve_is_bang (input)) || (use_backslash && verve_is_backslash (input)))
    {
      /* Launch DuckDuckGo */
      gchar* esc_input = g_strescape(input, NULL);
      command = g_strconcat ("verve-search-launcher https://duckduckgo.com/?q=", esc_input, NULL);
      g_free(esc_input);
    }
    else
    {
      if (use_smartbookmark)
      {
        /* Launch user-defined search engine */
        command = g_strconcat ("verve-search-launcher ", url, input, NULL);
      }
      else
      {
        /* Run command using the xfterm4 wrapper if the terminal flag was set */
        if (G_UNLIKELY (terminal))
          command = g_strconcat ("xfterm4 -e ", input, NULL);
        else
          command = g_strdup (input);
      }
    }
  }
    
  /* Try to execute the exo-open command */
  if (verve_spawn_command_line (command))
    result = TRUE;
    
  /* Free command string */
  g_free (command);

  /* Return spawn result */
  return result;
}



/*********************************************************************
 *
 * Internal pattern matching functions
 *
 *********************************************************************/

gboolean
verve_is_url (const gchar *str)
{
  GString     *string = g_string_new (str);
  pcre        *pattern;
  const gchar *error;
  int          error_offset;
  int          ovector[30];
  gboolean     success = FALSE;

  /* Compile first pattern */
  pattern = pcre_compile (MATCH_URL1, 0, &error, &error_offset, NULL);
  
  /* Test whether the string matches this pattern */
  if (pcre_exec (pattern, NULL, string->str, string->len, 0, 0, ovector, 30) >= 0)
    success = TRUE;

  /* Free pattern */
  pcre_free (pattern);

  /* Free data and return true if string matched the first pattern */
  if (success)
    {
      g_string_free (string, TRUE);
      return TRUE;
    }

  /* Compile second pattern */
  pattern = pcre_compile (MATCH_URL2, 0, &error, &error_offset, NULL);

  /* Test whether the string matches this pattern */
  if (pcre_exec (pattern, NULL, string->str, string->len, 0, 0, ovector, 30) >= 0)
    success = TRUE;

  /* Free pattern */
  pcre_free (pattern);

  /* Free string */
  g_string_free (string, TRUE);

  /* Return whether the string matched any of the URL patterns */
  return success;
}



gboolean
verve_is_directory (const gchar *str)
{
  /* Avoid opening directories with the same name as an existing executable. */
  if (g_find_program_in_path (str))
    return FALSE;

  if (g_file_test (str, G_FILE_TEST_IS_DIR))
    return TRUE;
  else
    return FALSE;
}



gboolean
verve_match_regex (const gchar *str, const gchar *regex)
{
  GString     *string = g_string_new (str);
  pcre        *pattern;
  const gchar *error;
  int          error_offset;
  int          ovector[30];
  gboolean     success = FALSE;

  /* Compile pattern */
  pattern = pcre_compile (regex, 0, &error, &error_offset, NULL);

  /* Test whether the string matches this pattern */
  if (pcre_exec (pattern, NULL, string->str, string->len, 0, 0, ovector, 30) >= 0)
    success = TRUE;

  /* Free pattern and string */
  pcre_free (pattern);
  g_string_free (string, TRUE);

  if (success)
    {
      /* Don't search DuckDuckGo for this query is there is an executable by this name. */
      if (g_find_program_in_path (str))
      {
        success = FALSE;
      }
    }

  /* Return true if string matched the first pattern and there is no matching executable */
  return success;
}



gboolean
verve_is_email (const gchar *str)
{
  return verve_match_regex(str, MATCH_EMAIL);
}



gboolean
verve_is_bang(const gchar *str)
{
  return verve_match_regex(str, MATCH_BANG);
}



gboolean
verve_is_backslash(const gchar *str)
{
  return verve_match_regex(str, MATCH_BACKSLASH);
}



/* vim:set expandtab sts=2 ts=2 sw=2: */
